### PR TITLE
Feature/attribute type selector

### PR DESCRIPTION
**Pull request by alexhunziker**
# Feature/attribute type selector

closes #na

## Description

Hey guys, since I am staying at home due to the virus this week i had some time and i thought it might be nice to switch the aggregation mode of metrics in the visualization on the fly. Therefore: here comes AttributeTypes III. This PR is dependent on the two pending 365 PRs.

- Attribute Types can now be selected in the visualization in the node/edge metric dropdowns

## Definition of Done

A task/pull request will not be considered to be complete until all these items can be checked off.

- [ ] **All** requirements mentioned in the issue are implemented
- [ ] Does match the Code of Conduct and the Contribution file
- [ ] Task has its own **GitHub issue** (something it is solving)
  - [ ] Issue number is **included in the commit messages** for traceability
- [ ] **Update the README.md** with any changes/additions made
- [ ] **Update the CHANGELOG.md** with any changes/additions made
- [ ] **Enough test coverage to ensure that uncovered changes do not break functionality**
- [ ] **All tests pass**
- [ ] **Descriptive pull request text**, answering:
  - What problem/issue are you fixing?
  - What does this PR implement and how?
- [ ] **Assign your PR to someone** for a code review
  - This person _will be contacted **first**_ if a bug is introduced into `master`
- [ ] **Manual testing** did not fail

### CODE REVIEW

**Comment by alschmut** [comment]
Thanks @alexhunziker for contributing so much to the never ending attributeTypes epic 😃 but I have to say, that, from a user perspective, I am not a fan of this PR. In my opinion the majority of all users should get the attributeTypes via the ccsh importers. And for everybody else, who might like to manually set the attributeType, we should rather think of one pop-up-dialog to set the attributeType for each metric with only one list of unique metrics. Having the option to set the attributeTypes in every node/edge-selector confused me and I also wouldn't know if I had to select it for every selector again..

I think for now we are pretty good with the other two PRs, when the ccsh importer deliver the attributeTypes (:

**Comment by alexhunziker** [comment]
Hey @alschmut thanks for the feedback on this one.

So why did I do this PR? I think the user should be able to select the aggregation mode on the fly because

1.  the ccsh will (with my PRs) only generate attribte Types for the tokei and scmlog importers. Depending on when someone (most likely not me; see below) feels like adapting the other importers, the correct AttributeTypes will not be available for many importers for quite some time
2. For many metrics both aggregation modes may be appropriate, so preselecting one during the ccsh file generation is restrictive and less than ideal.
3. In the future more aggregation modes (max/min/mean) could be introduced as a feature to the users.

For those reasons i think the functionality proposed in this PR is valuble.

Regarding the UX/design: I had two ideas on how to make this feature available to users; this one and something similar as you have proposed with a pop-up/single dropdow. I chose this option, since with the pop-up/separate dropdown there would be the need for another button in the already pretty crowded ribbon-bar which i wanted to avoid. However I am also not that convinced that my approach is the best from a UX perspective, so if someone has a better idea, feel free to change it.

Technically i tried to write the selector component in a way that the concept would also work if the selectors are placed/designed differently.

The bottom line is: I think this is a valuable feature (obviously, that is why I implemented it ;) ), but if someone has a better approach on how to make this feature available to the user, that is great too. As I am participating as an external contributor here, you are of course also free to decide to discard this PR entirely.

As I will have significantly less time starting from next week, I will probably not be able to contribute to codeCharta anymore, or at least significantly less frequently. Provided no major changes are requested, i will continue maintaining these 3 PRs until they are merged though. But the continuation of the AttributeTypes epic (provide AttributeTypes in the ccsh for all importers, optimizations in the visualization, ...) would thereafter up to you guys.

**Comment by BridgeAR** [comment]
Is it maybe possible to split this PR into multiple smaller ones? It's difficult to review a PR of this size.

**Comment by alschmut** [comment]
> Is it maybe possible to split this PR into multiple smaller ones? It's difficult to review a PR of this size.

@BridgeAR I think this PR is not actually as big, but also contains the commits of the two other PRs, as mentioned in the description. See the [comparison](https://github.com/MaibornWolff/codecharta/compare/feature/365/median-symbol...feature/attribute-type-selector) between this and the other PR branch (:

**Comment by BridgeAR** [comment]
@alschmut thanks for the hint.

**Comment by alexhunziker** [comment]
I think i found a way to make this feature better; I replaced the dots with the current attributeType/aggregation mode of the metric. This way the user gets feedback on changing the attribute type, and also already knows which aggregation mode will be used when selecting a metric.

aggregationSelector1
aggregationSelector2

**Comment by alschmut** [comment]
@alexhunziker thanks for the detailed information above! Actually I would accept your second idea, with showing the current symbol. This might in fact enhance the understanding of the attributeTypes for the user a lot. Just got some minor questions/ideas:
- On which side should the symbol be located? on the left of the metricName or on the right (where it is now)?
- Like with the nodeContextMenu, I would add the symbols also in the dropdown on the left of the attributeType name

**Comment by NearW** [comment]
This is a great solution! I really like it.

Can we get a screenshot of this on the left so that we can discuss this in our planning on friday?

File: visualization/app/codeCharta/ui/metricType/metricType.component.ts [hunk 1]
```diff
@@ -69,6 +73,14 @@ export class MetricTypeController
 	public onBuildingUnhovered() {
 		this._viewModel.isFolderHovered = false
 	}
+
+	public onMetricDataAdded() {
```
Comment (NearW) [inline visualization/app/codeCharta/ui/metricType/metricType.component.ts:77]: add test

File: visualization/app/codeCharta/ui/attributeTypeSelector/attributeTypeSelector.component.ts [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+	public setToRelative(metricName: string, category: string) {
+		this.setAttributeType(metricName, category, AttributeTypeValue.relative)
+	}
+
+	private setAttributeType(metricName: string, category: string, type: AttributeTypeValue) {
+		const attributeTypes = this.storeService.getState().fileSettings.attributeTypes
+		if (attributeTypes[category]) {
+			attributeTypes[category][metricName] = type
+		}
+		this.storeService.dispatch(setAttributeTypes(attributeTypes))
```
Comment (NearW) [inline visualization/app/codeCharta/ui/attributeTypeSelector/attributeTypeSelector.component.ts:42]: This conflicts with the immutable update pattern of redux. You're copying a reference of the attributeTypes and changing the store without dispatching an action (even though you do it later).

Instead define another action in `attributeType.actions.ts` that you can dispatch in order to change an attributeType by category and metricName.

`setAttributeTypeByName(category, metricName)` and then `this.storeService.dispatch(setAttributeTypeByName(category, metricName)`

Comment (alexhunziker) [reply visualization/app/codeCharta/ui/attributeTypeSelector/attributeTypeSelector.component.ts]: You are right, I hence created a new aciton for this. Not sure if the updateAttributeType function is in the right place within the reducer, if not pls move it to  a better place.

**Comment by NearW** [comment]
We decided that this feature will be merged to master, but before we will change the following things:
- Changing the attributeType of a metric will change the attributeType of the metric in the other metricChoosers as well
- Add an Icon in the dropdown as well
- Icons on the right

File: visualization/app/codeCharta/state/store/fileSettings/attributeTypes/attributeTypes.reducer.ts [hunk 3]
```diff
@@ -1,12 +1,22 @@
...
 			return clone(action.payload)
+		case AttributeTypesActions.UPDATE_ATTRIBUTE_TYPE:
+			return updateAttributeType(state, action)
 		default:
 			return state
 	}
 }
+
+function updateAttributeType(state: AttributeTypes, action: UpdateAttributeTypeAction) {
+	const copy = clone(state)
```
Comment (NearW) [inline visualization/app/codeCharta/state/store/fileSettings/attributeTypes/attributeTypes.reducer.ts:17]: use shallow copy instead

File: visualization/app/codeCharta/ui/metricType/metricType.component.spec.ts [hunk 4]
```diff
@@ -69,40 +69,70 @@ describe("MetricTypeController", () => {
...
-			expect(metricTypeController["_viewModel"].edgeMetricType).toBe(AttributeTypeValue.relative)
+			expect(metricTypeController["_viewModel"].metricType).toBe(AttributeTypeValue.relative)
+		})
+	})
+
+	describe("onMetricDataAdded", () => {
+		it("should update metricType for node selections", () => {
+			storeService.dispatch(setAttributeTypes({ nodes: {}, edges: { foo: AttributeTypeValue.relative } }))
+			metricTypeController["metricSelection"] = MetricSelections.heightMetric
+			storeService.getState = jest.fn().mockReturnValue({ dynamicSettings: { heightMetric: "foo" } })
```
Comment (NearW) [inline visualization/app/codeCharta/ui/metricType/metricType.component.spec.ts:117]: store.dispatch(setHeightMetric)

File: visualization/app/codeCharta/ui/metricType/metricType.component.spec.ts [hunk 5]
```diff
@@ -69,40 +69,70 @@ describe("MetricTypeController", () => {
...
+			metricTypeController.onMetricDataAdded()
+
+			expect(metricService.getAttributeTypeByMetric).toBeCalledWith("foo")
+			expect(metricTypeController["_viewModel"].metricType).toBe(AttributeTypeValue.relative)
+		})
+
+		it("should update metricType for edge selection", () => {
+			storeService.dispatch(setAttributeTypes({ nodes: {}, edges: { foo: AttributeTypeValue.relative } }))
+			metricTypeController["metricSelection"] = MetricSelections.edgeMetric
+			storeService.getState = jest.fn().mockReturnValue({ dynamicSettings: { edgeMetric: "foo" } })
```
Comment (NearW) [inline visualization/app/codeCharta/ui/metricType/metricType.component.spec.ts:129]: storeService.dispatch(setEdgeMetric)

File: visualization/app/codeCharta/ui/attributeTypeSelector/attributeTypeSelector.component.spec.ts [hunk 6]
```diff
@@ -32,32 +32,21 @@ describe("AttributeTypeSelectorController", () => {
...
-		storeService.getState = () => {
-			return ({ fileSettings: { attributeTypes: { nodes: { bar: "relative" }, edges: {} } } } as unknown) as State
-		}
 	}
 
 	describe("setToAbsolute", () => {
-		it("should change attributeType to absolute", () => {
-			const expected = setAttributeTypes({ nodes: { bar: AttributeTypeValue.absolute }, edges: {} })
+		it("should update attributeType to absolute", () => {
+			const expected = updateAttributeType("nodes", "bar", AttributeTypeValue.absolute)
```
Comment (NearW) [inline visualization/app/codeCharta/ui/attributeTypeSelector/attributeTypeSelector.component.spec.ts:39]: use storeService to arrange data?
